### PR TITLE
Add an example of a WAL decoder

### PR DESCRIPTION
**Pull request by daamien**
Hi !

I recently wrote this example as an exercice to learn how to build a WAL decoder with PGRX.... I built this as « the thing I would have loved to have before » when I started digging on this topic. Feel free to modify it heavily or discard it completely if it is too specific or if the coding style is not up to the PGRX standards. I tried my best to write idiomic Rust but there may be some naïve/unsound parts...

This example shows how to build a basic Change Data Capture (CDC) mechanism using the Postgres Logical Decoding capabilities. The changes occurring on the database are serialized into JSON and pushed to a queue (a "logical replication slot") where they can be consumed by remote clients.

A Postgres CDC extension can be a used in various purposes:

* Ad hoc replication systems (e.g. Postgres => SQL Server )
* External commit-log for a distributed system (such as Kafka)
* Advanced Monitoring ( e.g. Prometheus/Loki )

This example tries to find the right tradeoff between simplicity and usefulness. Currently it has strong limitations but they should be easy to overcome.

Rust really shines in this example compared to similar implementations in C (see the wal2json extension). The Serde crate provides JSON serialization out of the box, whereas all the C implementations are forced to write their own JSON formatter.

### CODE REVIEW

**Comment by eeeebbbbrrrr** [comment]
Could I trouble you to put a README.md together for this crate that details what it does and how to use it?

**Comment by daamien** [comment]
> Could I trouble you to put a README.md together for this crate that details what it does and how to use it?

Sure, I wrote the explanations in the preamble comment of the `src/lib.rs` but I can move it into README file

File: pgrx-examples/decoder/src/lib.rs [hunk 1]
```diff
@@ -240,10 +240,8 @@ unsafe extern "C" fn pg_decode_begin_txn(
     ctx_ptr: *mut pg_sys::LogicalDecodingContext,
     _txn_ptr: *mut pg_sys::ReorderBufferTXN
 ){
-    let ctx = unsafe { PgBox::from_pg(ctx_ptr) } ;
-    let mut state = unsafe {
-        PgBox::from_pg(ctx.output_plugin_private as *mut DecodingState)
-    };
+    let ctx = PgBox::from_pg(ctx_ptr);
+    let mut state = PgBox::from_pg(ctx.output_plugin_private as *mut DecodingState);
```
Comment (workingjubilee) [inline pgrx-examples/decoder/src/lib.rs]: You're gonna want to revert this commit, because this should have `deny(unsafe_op_in_unsafe_fn)`.

File: pgrx-examples/decoder/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,322 @@
+use pgrx::prelude::*;
```
Comment (workingjubilee) [inline pgrx-examples/decoder/src/lib.rs]:
```suggestion
#![deny(unsafe_op_in_unsafe_fn)]
use pgrx::prelude::*;
```

File: pgrx-examples/decoder/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+mod tests {
+    use pgrx::prelude::*;
+
+    #[pg_test]
+    #[ignore]
+    fn test_decoder() {
+        //
+        // Not sure how to handle the tests of this extension....
+        // This function is running *inside* a transaction
+        // and the WAL decoder will output the changes only when they ar
```
Comment (workingjubilee) [inline pgrx-examples/decoder/src/lib.rs]:
```suggestion
        // and the WAL decoder will output the changes only when they are
```

File: pgrx-examples/decoder/src/bin/pgrx_embed.rs [hunk 4]
```diff
@@ -0,0 +1 @@
+::pgrx::pgrx_embed!();
```
Comment (workingjubilee) [inline pgrx-examples/decoder/src/bin/pgrx_embed.rs]: nit but I just don't like GitHub whining at me about it.
```suggestion
::pgrx::pgrx_embed!();

```

File: pgrx-examples/decoder/README.md [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+* They build Rust structs (Action, Tuple) to represent those changes
+* The structs are then serialized into JSON
+* The JSON output is sent into a replication slot
+* The ouput be can consumed in various ways by a remote client.
+
+
+## Requirements
+
+In order to use this extension with a cargo-pgrx managed database, you'll
+need to add the configuration below in "$PGRX_HOME/data-$PGVER/postgresql.conf".
```
Comment (workingjubilee) [inline pgrx-examples/decoder/README.md]: isn't this "any" database, to the appropriate postgresql.conf?

File: pgrx-examples/decoder/README.md [hunk 6]
```diff
@@ -0,0 +1,119 @@
+
+# A simple Change Data Capture ([CDC]) extension.
+
+This extension will extract the [DML] changes from Postgres WAL
+using [Logical Decoding] and export them in JSON format using [serde].
+
+## Principles
+
+* Postgres triggers various callbacks at the different stages of a transaction
+* The decoder defines some of these callbacks : begin, change, commit, etc.
```
Comment (workingjubilee) [inline pgrx-examples/decoder/README.md]:
```suggestion
* The decoder defines some of these callbacks: begin, change, commit, etc.
```

File: pgrx-examples/decoder/README.md [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+* The callbacks extract the changes made during the transaction
+* They build Rust structs (Action, Tuple) to represent those changes
+* The structs are then serialized into JSON
+* The JSON output is sent into a replication slot
+* The ouput be can consumed in various ways by a remote client.
```
Comment (workingjubilee) [inline pgrx-examples/decoder/README.md]:
```suggestion
* The output can be consumed in various ways by a remote client
```

File: pgrx-examples/decoder/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+    #[must_use]
+    pub fn postgresql_conf_options() -> Vec<&'static str> {
+        // return any postgresql.conf settings that are required for your tests
+        vec![]
+    }
+}
+
+// <!-- EOL -->
+
+
```
Comment (workingjubilee) [inline pgrx-examples/decoder/src/lib.rs]:
```suggestion
```

File: pgrx-examples/decoder/README.md [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+     |     "change_count": 1            +
+     | }
+```
+
+## Limitations
+
+This decoder is designed as a basic example and it has the following limitations:
+
+* Only the REPLICA IDENTITY FULL mode is fully supported. Supporting REPLICA IDENTITY DEFAULT
+  would require additionnal work.
```
Comment (workingjubilee) [inline pgrx-examples/decoder/README.md]:
```suggestion
  would require additional work.
```

File: pgrx-examples/decoder/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+    committed: Option<i64>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    rel: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    old: Option<Tuple>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    new: Option<Tuple>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    change_count: Option<i64>,
+}
```
Comment (workingjubilee) [inline pgrx-examples/decoder/src/lib.rs]: Can you add a unit test to this crate that attempts to serialize this struct, with all the Options as None, and then deserialize this struct from the resulting string?

File: pgrx-examples/decoder/src/lib.rs
Comment (workingjubilee) [inline pgrx-examples/decoder/src/lib.rs:1]: Could you run `cargo fmt` on this? Thank you.

**Review by workingjubilee — CHANGES_REQUESTED** [review]
Thanks! Looks good to me. I indeed have some mostly-stylistic nits in mind here.

File: pgrx-examples/decoder/README.md (hunk 5 above)
Comment (daamien) [reply pgrx-examples/decoder/README.md]: I think I meant "cargo-pgrx managed instance"

**Comment by workingjubilee** [comment]
I'm perplexed why we don't implement both deserialization and serialization.

**Comment by daamien** [comment]
> I'm perplexed why we don't implement both deserialization and serialization.

The purpose of a WAL decoding extension is to receive a WAL through the logical decoding mechanism and transform it into text representations of the operations performed. In this example, the extension will produce a JSON output but it will never have to injest any JSON input. It's really a one way road.

The question of who will consume the JSON output and how it will be consumed is out of the scope.

Imagine that you are legally required to archive any change made to a given table in your database. You might want to use a Kafka storage for that.  First the WAL decoder will generate the JSON output in the replication slot and then you'd write a client that consumes events in the replication slot and push them to kafka.

**Comment by workingjubilee** [comment]
Hmm. I guess it just feels weird that we don't check we can reconstruct our type from its own output. If we could, that capability... the interface type... could live in its own library, you see, that both producer and consumer could reference.

File: pgrx-examples/decoder/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,319 @@
...
+mod tests {
+    use pgrx::prelude::*;
+
+    #[pg_test]
+    fn test_action_begin() {
+        use crate::Action;
+
+        let input = Action::begin();
+        let json = serde_json::to_string(&input).expect("Serde Error");
+        assert_eq!(json, "{\"typ\":\"BEGIN\"}");
```
Comment (workingjubilee) [inline pgrx-examples/decoder/src/lib.rs]: We don't necessarily need vigorous escaping if we use a raw string:
```suggestion
        assert_eq!(json, r#"{"typ":"BEGIN"}"#);
```

**Review by workingjubilee — APPROVED** [review]
I think this looks fine.

**Comment by workingjubilee** [comment]
Does it work with `./run-examples.sh`?

File: pgrx-examples/decoder/Cargo.toml [hunk 12]
```diff
@@ -0,0 +1,38 @@
+[package]
+name = "decoder"
```
Comment (workingjubilee) [inline pgrx-examples/decoder/Cargo.toml]: oh, only last thought is that it should probably be named `wal_decoder` instead? likelihood of another `decoder` example appearing over sufficient time is approximately 1

Comment (daamien) [reply pgrx-examples/decoder/Cargo.toml]: agreed.

Maybe wal2json_decoder ? There may be others types of WAL decoder in the future ?

**Comment by daamien** [comment]
> Does it work with `./run-examples.sh`?

It does :)

**Comment by daamien** [comment]
Renamed , Squashed and Rebased :)

**Comment by workingjubilee** [comment]
Thanks!

Oh, re: soundness, nothing seems obviously wrong? But I haven't mind-numbingly-closely audited it for soundness as it's just an example, and thus is user code, basically: any brain cells I spend on doing that are better off spent on making it easier to write code soundly than torturing you with revisions of some pretty gnarly unsafe code.
